Doc: ogr2ogr: Clarify use of -select, -fieldmap

simonsouth opened the pull request:
## What does this PR do?

Extend `ogr2ogr`'s documentation to clarify the use of the "-select" and "-fieldmap" options. This would have saved me time this morning when I was trying to get `ogr2ogr` to omit a foreign-key column from a layer being loaded from a GeoPackage.

As I eventually discovered,

- [The "-select" option is quietly ignored when "-append" is specified.](https://github.com/OSGeo/gdal/blob/d1f5a1634136677bacd04fc6e07d2673b42d0f1c/gdal/apps/ogr2ogr_lib.cpp#L3394)

- "-fieldmap" can be used instead to omit source fields by specifying a value of -1 for their target index, [as shown in the last example on the page](https://www.gdal.org/ogr2ogr.html#ogr2ogr_example).

Since it's a little hard to tell from the diff, the specific text added to the page is:

"Note this setting is ignored when -append is used. To control the selection of fields when appending to a layer, use the -fieldmap setting."

"To omit a field, specify a value of -1."

## What are related issues/pull requests?

None.

## Tasklist

 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

rouault commented:
As people don't read docs :-), it would also be nice if you could make ogr2ogr emit an error if -select is used in a -append context.

Note that for your use case, you can also use -sql "SELECT src_field_name_1 [AS dst_field_name1], .... FROM source_layer_name"

simonsouth commented:
I've added a commit that causes ogr2ogr to exit if "-select" and "-append" are specified together. Let me know if this is what you had in mind.

rouault commented:
hum, this is interesting. The failure of ogr2ogr_lib_19 shows that actually -select is taken into account at layer creation of if -addfields is specified. So the test and doc should be refined

simonsouth commented:
I'll have to take a closer look at this a bit later. Probably the best thing to do next is to write a test case that demonstrates the problem I was seeing this morning.

simonsouth commented:
I've added a test case that demonstrates the (to me) surprising behaviour of ogr2ogr when the "-append" and "-select" options are specified together. You can run it with

```
cd autotest/utilities
python ./test_ogr2ogr_lib.py
```

[The code](https://github.com/OSGeo/gdal/blob/a7b71eb4c9a1e944d94623331914a6de4895db1c/autotest/utilities/test_ogr2ogr_lib.py#L504-L541) is based on [the `ogr2ogr_lib_19` test case](https://github.com/OSGeo/gdal/blob/a7b71eb4c9a1e944d94623331914a6de4895db1c/autotest/utilities/test_ogr2ogr_lib.py#L433-L463) that failed previously. It

- [Creates a feature](https://github.com/OSGeo/gdal/pull/864/commits/a7b71eb4c9a1e944d94623331914a6de4895db1c#diff-50a2d94c5aa77bcbfe61960f194c23baR511) with two properties, "foo" and "bar";
- [Adds the feature as-is](https://github.com/OSGeo/gdal/pull/864/commits/a7b71eb4c9a1e944d94623331914a6de4895db1c#diff-50a2d94c5aa77bcbfe61960f194c23baR518) to a new layer in memory; and
- [Appends the same feature to the layer](https://github.com/OSGeo/gdal/pull/864/commits/a7b71eb4c9a1e944d94623331914a6de4895db1c#diff-50a2d94c5aa77bcbfe61960f194c23baR519) again, using the equivalent of the "-select" option to specify only the feature's "foo" field should be included.

I expected the result to be that the second feature would contain only a "foo" field, since that was the one field selected for inclusion.

Instead both features contain _both_ fields&mdash;and not only does the second feature contain a "bar" field, it is set to its original value! So while I expected the "-select" option to function like a projection operator, it seems to do _nothing at all_ when the "-append" option is included.

Does this strike you also as surprising behaviour, or am I simply not understanding correctly what the "-select" option is meant to do?

rouault commented:
> Does this strike you also as surprising behaviour, or am I simply not understanding correctly what the "-select" option is meant to do?

Yes, that's definitely surprising. So either we error out in that case, or we implement it properly.
-append -sql "SELECT foo FROM source" does work properly.

simonsouth commented:
> So either **we error out in that case**, or we implement it properly.

I've updated the pull request to reflect the former approach. This is because I feel [lines 3790](https://github.com/OSGeo/gdal/blob/master/gdal/apps/ogr2ogr_lib.cpp#L3790) and [3815](https://github.com/OSGeo/gdal/blob/master/gdal/apps/ogr2ogr_lib.cpp#L3815) of [ogr2ogr_lib.cpp](https://github.com/OSGeo/gdal/blob/master/gdal/apps/ogr2ogr_lib.cpp
) make it clear the original intent was for the -append option, when specified, to be used with -fieldmap and not -select:

```
if (m_papszFieldMap && bAppend) ...
else if (m_papszSelFields && !bAppend ) ...
```

Given unlimited time and resources I'd take a shot at refactoring ogr2ogr to function more like a transformation pipeline, with (for instance) the -sql, -select and -fieldmap options applied in every case and in that specific order. But I have other priorities to attend to right now.

The updated pull request

- Causes ogr2ogr to fail with an error message if -select and -append (but not -addfields) are specified together. I've implemented this in a different spot in ogr2ogr_lib.cpp that places it alongside other, existing tests.

- Updates the corresponding test case to verify the library returns an "invalid argument" error in the case -select and -append are specified together.

- Updates the documentation for ogr2ogr as before, except to make it clear the tool explicitly does not support -select and -append being used together and to mention the -sql option as another alternative.
